python/ray/data/_internal/logical/interfaces/plan.py

```diff
@@ -8,6 +13,13 @@ class Plan:
     that's useful for plan optimization or execution.
     """
 
+    def __init__(self):
+        self._context: Optional["DataContext"] = None
+
     @property
     def dag(self) -> Operator:
         raise NotImplementedError
+
+    def _set_context(self, context: Optional["DataContext"]):
```

Comment:
Nit: Could potentially use `@property` here but not sure if that really gets you much.